### ISSUE TITLE
SVG 3

### ISSUE THREAD

**Issue by CodyKarch** [ebeshero/DHClass-Hub#78]
`select="./f[@select='Yes']/@n / sum(@n) * 100"`
Here is what I'm trying to use to represent n divided by the sum of n times 100, but I know this is wrong, and help? @ebeshero @RJP43

**Comment by spadafour** [comment]
I'm not sitting in front of my computer, but it might help to think about this in 3 steps. First, we need to figure out the number of times that someone answered yes to this question(or no, yes but fined, or blank, depending on what bar you are working on). Then, you need to figure out the percentage of those answers by totaling the number of  answers for that question and dividing by the total yes, no, yes but fined, and blank answers. I wrote a separate rule for each response to do this. The issue with this is that some answers have yes but fined, but some don't. To get around this, I wrote an xsl:choose (when one of the possible answers for the question does contain yes but fined as an answer, add the number of yes but fined responses to the denominator. Otherwise (if there is no yes but fined answer), the denominator is just the number of yes plus the number of no plus the blank).

So what do we have so far? We have 4 rules: one for yes, one for no, one for yes but fined, and one for blank. These all (except for the yes but fined bar) have xsl:choose statements with 2 options, 1 that creates a percentage if yes but fined is an option, and another that creates a percentage if yes but fined is not an option.

I never multiply by 100 in my homework. Why?Think about a test: if you score 80 out of 100 points on a test, you get 0.80, or 80%. If you score 40/50, you get the same score. Multiplying by 100 just makes your grade easier to read and understand. Getting an 80% means that you got 0.80 of the Total possible points. To put it in math terms, you got:
0.8 TIMES the total number of questions.
So, with all that said:

The number that we created with our rules should be a decimal value. We need to multiply that by the height of our y-axis (why? Because our y axis represents 100%, or, the top of our y axis is the total possible amount of points/responses that a question can get).

Basically, this is a lot of math and thinking about each response in terms of a percent of your y-axis.

I hope some of this helps, I'm typing this on a phone on my way to an airport. I'll try and help further when I land/if I have time waiting for my flight.

Once you understand what we are doing conceptually (which is mainly what I tried to explain), you can then firgure out the specifics of grabbing all of the numbers and making formulas.

**Comment by spadafour** [comment]
Also: I used xsl:choose to get around the problem where yes but fined may or may not be a response, but there is probably a more elegant way to do this (my guess is with sum(), but I didn't use that)

**Comment by spadafour** [comment]
Troubleshooting on the fly: is sum(@n) actually getting you a sum of ALL possible responses, or just the responses for select=yes? Also, '/' doesn't mean divide in Xslt, the code for that is 'div'

**Comment by CodyKarch** [comment]
Wow. That cleared a lot up for me, Thank you very much. Now I just need to put it into practice.

**Comment by CodyKarch** [comment]
`<xsl:variable name="spots" select="position() * $Interval"/>`
`<xsl:variable name="Yes" select="((./f[@select='Yes']/@n) div (./@n)) * (100)"/>`
`<line x1="{$spots}" y1="0" x2="{$spots}" y2="-{$Yes}" color="black" stroke="black" stroke-width="3"/>`
`<xsl:variable name="Nos" select="((./f[@select='No']/@n) div (./@n)) * (100)"/>`
`<line x1="{$spots}" y1="0" x2="{$spots}" y2="-{$Nos}" color="black" stroke="black" stroke-width="3"/>`
This is what I have, but I know it isn't fully right. Any clue?

**Comment by brookestewart** [comment]
I'm having an issue with the same step. It will give a number for `<xsl:template match="//f[@select='Yes']">
        <xsl:variable name="xPos" select="position()*$Width"/>
            <xsl:variable name="yesPerc" select="(//f[@select='Yes']/@n)"/>
            <line x1="{$xPos}" y1="0" x2="{$xPos}" y2="-{$yesPerc}" stroke="blue" stroke-width="30"/>
    </xsl:template>`
But it gives and error when I try to calculate the length: 
`<xsl:variable name="yesPerc" select="(((//f[@select='Yes']/@n) div sum(@n)) * 100)"/>` it says "A sequence of more than one item is not allowed as the first operand of 'div'" Can anyone help me?

**Comment by ebeshero** [comment]
@brookestewart I'm worried about your positioning of those forward slashes, and the node you've selected for your template match. Remember why we use the dog notation (.) in XSLT! I think that might be causing the problem here. 

Think about your template match. You should be able to plot all your bar-lines (for yes, no, blank, etc) all in one and the same template rule, and you need to do that because their positions on the graph each depend on each other. Hope that helps!

**Comment by ebeshero** [comment]
@CodyKarch I jus returned home and don't have my computer out, but I think from eye-balling your code that you are likely having a problem with your last-drawn bar (for Nos) coveting up your Yes bar, because they're both occupying the same coordinate space: you've drawn each one to start from zero! But don't you want the Nos to be plotted on top of the Yeses? You need start your y1 for Nos at the same point as the y2 for Yeses.

**Comment by CodyKarch** [comment]
I do see what you mean for that, but my main problem is the spacing between bars and how I'm not getting output.

**Comment by ebeshero** [comment]
@CodyKarch are you having trouble with the X positions? If so, that's going to be a question of the context in which you are defining your position variables and lines. Are you doing this within XSL:for-each? Or a template match? what are you matching on, if it's a template match? Or, what are you plugging in for your @select on XSL:for-each?

**Comment by brookestewart** [comment]
@ebeshero Thanks, that makes sense! I changed it to: 
`<xsl:template match="//fs[f[@select='Yes']]">` 
        `<xsl:variable name="xPos" select="position()*$Width"/>`
       `<xsl:variable name="yesPerc" select="(((./f[@select='Yes']/@n) div sum(@n)) * 100)"/>`
            `<line x1="{$xPos}" y1="0" x2="{$xPos}" y2="-{$yesPerc}" stroke="blue" stroke-width="30"/>`
    `</xsl:template>` 
but it still won't work. I'm not sure what's wrong now.

**Comment by ebeshero** [comment]
@brookestewart OK--think about your position() function. Do you need to indicate that this is the position of the current context node?

**Comment by brookestewart** [comment]
I tried `position(.)` but it's giving me an error. I'm so lost right now.

**Comment by brookestewart** [comment]
This is what one of the output lines looks like, if that helps.
`<line x1="480"
            y1="0"
            x2="480"
            y2="-INF"
            stroke="blue"
            stroke-width="30"/>`

**Comment by ebeshero** [comment]
@brookestewart Sorry--I was asleep! But as you probably figured, it looks like your y2 is calculation something infinite, which is what happens when you divide by zero or by something that isn't there. The problem is the XPath expressions you're using to reach the numbers you want--something isn't reaching where you think it is. You are dividing by just the `sum(@n)` to get all the values of `@n` on the `<f>` elements *inside* the `<fs>`. But your XPath is looking for an `@n` on the current context node, `<fs>`, where there isn't an `@n`, and the result is infinity because you wind up dividing by zero. 

Position() as it is might be working fine now that you have redefined your context. You want the position() to be based on the question.

**Comment by ebeshero** [comment]
Though this probably didn't cause the infinity problem in your y2, there is no need for those leading slashes in your template `@match` which only looks for a pattern wherever it sits in the XML hierarchy.
